### ISSUE TITLE
Allocation of Emissions Page

### ISSUE THREAD

**Issue by patriciarussellCAS** [bcgov/cas-reporting#287]
#### Description:
Blocked by #203 
Regulated Operations will need to allocate their emissions to products. This will happen after they have entered production data. 

Note for devs: Note devs: If a facility says they only produce one product on the earlier screen, maybe the allocation could auto-populate? (i.e., if there's only one product, 100% of emissions get allocated to that product)

Note2: [Flagging that there may be sources that do NOT get allocated ](https://www.figma.com/design/LsDmLDCdnJqI8UahyqvOD8?node-id=13583-3949#816451456)

[Wireframe](https://www.figma.com/design/LsDmLDCdnJqI8UahyqvOD8/BCIERS-Designs?node-id=13583-3949&t=r8pWMkO2XGlzxHSo-4)

Note: the ```Total allocated``` field should display as a percentage to 4 digit precision

#### Acceptance Criteria:

Given that I am a Regulated Operation
When I've completed the emissions and production data pages
Then I want to see the allocation of emissions page so that I can fill it out for the facility

Given that I am completed the emissions and production data pages
When I am on the allocation page
Then I can enter production allocation values for each reported product for each emission category

Given that I am completed the emissions and production data pages
When I am on the allocation page
Then I can see the total attributable emissions at the bottom of the page

Given that I am completed the emissions and production data pages
When I am on the allocation page and I have entered allocation data
Then I can see the total attributed emissions to each product

#### Development Checklist:
**BE**
- [ ] Create TimeStampedModel `report_product_emission_allocation` table {id, report_version,facility_report, report_product_id, emission_category_id, allocated_quantity} (`ReportProductEmissionAllocation`)
- [x] In  model  `report_product_emission_allocation`  set the foreign key relationships related_name=related_name="%(class)s_records",
- [x] In test files of the foreign key relationships  add `ReportProductEmissionAllocation` reference:
```
        cls.field_data = [...   
  ("reportproductemissionallocation_records", "report product emission allocation", None, 0),
       
     ]
 - `bc_obps/reporting/tests/models/test_emission_category.py` 
 -  `bc_obps/reporting/tests/models/test_facility_report.py`
 -  `bc_obps/reporting/tests/models/test_report_product.py` 
 -  `bc_obps/reporting/tests/models/test_report_version.py 
 - 
```
- [ ] In file `bc_obps/registration/tests/models/test_user.py` add  `ReportProductEmissionAllocation` timestamp field references:
```
            ("reportproductemissionallocation_created", "report product emission allocation", None, None),
            ("reportproductemissionallocation_updated", "report product emission allocation", None, None),
            ("reportproductemissionallocation_archived", "report product emission allocation", None, None),
       
```
- [x] In file `bc_obps/reporting/models/__init__.py` add `ReportProductEmissionAllocation` references:
```
from .report_product_emission_allocation import ReportProductEmissionAllocation

__all__ = [...
    "ReportProductEmissionAllocation"
]
```
- [x] Create model migration:
 ```
cd bc_obps
make migrations
make reset_db
```

- [x] POST API route/service to save allocation of emissions data to the `report_product_emission_allocation` table
- [x] GET API route/service to sum the emissions in the report by emission category by product (dependent on #333)
  Returns the emission categories `to allocate` across products for this report version\facility (categories schema)
  Returns the reported emission allocation `allocated` across products (initialData)
- [x] Add API routes to `bc_obps/reporting/api/__init__.py`
- [ ] Update file `make pythontests ARGS='common/tests/endpoints/auth/test_endpoint_permissions.py'` with newly created APIs

**FE**
- [x] Methodology selection {Calculator, Other} is a dropdown & if "other" is selected then a description field is shown
- [ ] There is a field under each emission category for each product reported 
- [ ] Each category product name and product emission fields display in 1 row  with 2 column
- [x] Shows the total attributable emissions and the totals allocated at the bottom of the page
- [ ] Validation: the sum of allocated emissions to products within a category cannot exceed the total emissions for that category
- [x] Validation: the sum of all allocated emissions to products cannot exceed the total attributable emissions
- [x] Saves the data to the report_product_emission_allocation table on clicking 'save & continue'
- [ ] Meets the DOD

**Definition of Ready** (Note: If any of these points are not applicable, mark N/A)

- [ ] User story is included
- [ ] User role and type are identified
- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if required)
- [ ] Design / Solution is accepted by Product Owner
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

**Definition of Done** (Note: If any of these points are not applicable, mark N/A)

- [ ] Acceptance criteria are tested by the CI pipeline
- [ ] UI meets accessibility requirements
- [ ] Configuration changes are documented, documentation and designs are updated
- [ ] Passes code peer-review
- [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
- [ ] Ticket is ready to be merged to main branch
- [ ] Can be demoed in Sprint Review
- [ ] Bugs or future work cards are identified and created
- [ ] Reviewed and approved by Product Owner

#### Blocked By
#203 
#333 

#### Notes:

-

**Comment by dleard** [comment]
@patriciarussellCAS @suhafa @pbastia blocked until we receive the "rules" about how to apply emission categories based on the data we collect. Adding the attn: Business Area tag

**Comment by patriciarussellCAS** [comment]
@dleard  confirming from BA that we need allocate emissions in all emissions categories, not just the basic ones.

**Comment by patriciarussellCAS** [comment]
There have been some updates after today's Giraffe collab and the meeting on FOG and Kiln emissions. We'll need to update the wireframe to reflect these changes, but some items are still being discussed by the BA. @zoeyli-46 

@acatchpole - just flagging that we are still confirming the final decisions on this form so it's still blocked. Sorry about that! 
@pbastia @dleard

**Comment by patriciarussellCAS** [comment]
As mentioned in the teams channel - our AC is incorrect. This will need to be updated: 
Given that I am an LFO Regulated Operations
When I have cycled through the emissions and production data for each facility
Then I want to see the the allocation of emissions page so that I can fill it out for each facility.

**Comment by patriciarussellCAS** [comment]
Flagging for developers that I have created a follow-up ticket to deal with the specific conditions related to this page. 
For reference: #[412](https://github.com/bcgov/cas-reporting/issues/412)

**Comment by acatchpole** [comment]
per Dev discussions with @dleard :

Q: Should ```report_version_id``` be included included in the ```report_product_emission_allocation``` model?
A: Yes, even though it is duplicating data, it should make queries easier to perform/understand later.

Q: Do we have a rule for what we set ```related_name``` to when defining foreign keys in the model?
A: tying it to the class name seems to be our convention (```"%(class)s_records"```)

Q: What all emission categories are we allocating to?
A: There are 3 emission category Types : Basic, fuel_excluded, and other_excluded. Total emissions are attributed to the basic categories. Excluded emissions can be attributed to the fuel_excluded type categories. Allocating emissions to other_excluded categories has been removed as a requirement by the BA.

Q: Does the total emissions allocated calculation include excluded emissions?
A: No. This total should reflect the total report attributable emissions which is all emissions attributed under basic categories only. The total attributable amount is included in the emission_category_service Get under the ```attributable_for_reporting``` property.

Q: Is there any prior examples of dynamically setting the label of a field along side its value? ProductName : allocated_amount
A: Possibly. Can't find one. Likely solutions are either:
   - Dynamically generate the schema for the product arrays with the label set then
   - Pass the value down through template props. Will probably need to create a new template to support this particular form.
